### DISCUSSION TITLE
Is there a way to point to a glob of c files to index automatically ?

### DISCUSSION THREAD

**Discussion by Strackeror** [clangd/clangd#1527] [Q&A] [answered]
Hello, I'm experimenting with a c repo that's structured in a way that makes generating a complete `compile_commands.json` a bit annoying (https://github.com/pret/pokeemerald).

I wrote a small `compile_flags.txt` that makes the opened files work pretty well in vscode. The only issue is that go to definition, find references and switch header/source only work with c files that have been opened at least once. Is there any fast way to point to a list of c files to 'open' in that way ? Or should I just bite the bullet and generate the `compile_commands.json` with some scripting ?

**Comment by HighCommander4** [comment, marked as answer]
Clangd relies on the list of entries in `compile_commands.json` to know which files to index (and the index is what powers the ability to find definitions/references in other translations units).

It looks like this project has a makefile -- for such projects, the easiest way to generate a `compile_commands.json` file is to use a tool like [Bear](https://github.com/rizsotto/Bear) which wraps a build command like `make`, detects what compiler commands it invokes, and records them.

**Reply by Strackeror** [reply]
The toolchain is very... nonstandard and the codebase makes heavy use of non-standalone headers, so automatic tools can only help up to a point here.

I was just hopeful there was an alternative to compile_commands.json for that purpose, but if there isn't I'll just take the time to write a script to generate it.

Nonetheless, thank you for taking the time to answer !

**Reply by i-ky** [reply]
> The toolchain is very... nonstandard and the codebase makes heavy use of non-standalone headers, so automatic tools can only help up to a point here.

Non-standalone headers shouldn't be a problem, clangd processes header file in the context of the source file that includes it.

As for nonstandard toolchain, I am in a similar situation. I would recommend trying out [compile-db-gen](https://github.com/sunlin7/compile-db-gen) if you are on a system with `strace` available and Bear does not work for you for some reason. I am also working on [a `compile_commands.json` generator that uses `ptrace` directly](https://github.com/i-ky/basset). It is still work in progress, but I would be glad to include your project as a test case.

**Reply by Strackeror** [reply]
> on-standalone headers shouldn't be a problem, clangd processes header file in the context of the source file that includes it.

It... does ? The documentation is very clear that headers have to be self contained to work. And even after generating the compilation database, I can confirm they don't for this project

> Commands for header files
> [#](https://clangd.llvm.org/design/compile-commands#commands-for-header-files)
> Clangd parses headers like any other source file (which is why it [only supports self-contained headers](https://github.com/clangd/clangd/issues/45)). However most build systems don’t compile headers directly and therefore don’t record compile commands for them.

**Reply by HighCommander4** [reply]
I think what @i-ky meant is that the presence of non-standalone headers should not impact clangd's ability to parse the (standalone) files that include them.

Indeed opening the non-standalone header itself won't work well, but that seems like an orthogonal problem to how the `compile_commands.json` file is generated.

**Reply by i-ky** [reply]
Or just stuff that I read about how clangd, ccls and Microsoft's extension work blended in my head... :)
